tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -3844,6 +3844,263 @@ LogicalResult lowerWithScheduleSparseReshapeOpCPU(
+    return dominant_op->emitError()
+           << "root_ops[0] is not a lmhlo_disc::SparseFillEmptyRowsOp";
+  }
+  auto sparse_fill_empty_rows_op =
+      cast<lmhlo_disc::SparseFillEmptyRowsOp>(root_ops[0]);
+  if (!sparse_fill_empty_rows_op) {
+    return dominant_op->emitError()
+           << "can not cast root_ops[0] to lmhlo_disc::SparseFillEmptyRowsOp";
+  }
+
+  auto loc = sparse_fill_empty_rows_op.getLoc();
+  OpBuilder b(root_ops.back());
+
+  b.setInsertionPoint(sparse_fill_empty_rows_op.getOperation());
+
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+  Value false_value =
+      b.create<arith::ConstantOp>(loc, b.getIntegerAttr(b.getI1Type(), false));
+  Value true_value =
+      b.create<arith::ConstantOp>(loc, b.getIntegerAttr(b.getI1Type(), true));
+  Value zero_int = b.create<arith::ConstantOp>(
+      loc, b.getIntegerAttr(b.getIntegerType(64), 0));
+  Value one_int = b.create<arith::ConstantOp>(
+      loc, b.getIntegerAttr(b.getIntegerType(64), 1));
+
+  // input
+  Value indices = sparse_fill_empty_rows_op.indices();
+  Value values = sparse_fill_empty_rows_op.values();
+  Value dense_shape = sparse_fill_empty_rows_op.dense_shape();
+  // output
+  Value output_indices = sparse_fill_empty_rows_op.output_indices();
+  Value output_values = sparse_fill_empty_rows_op.output_values();
+  Value empty_row_indicator = sparse_fill_empty_rows_op.empty_row_indicator();
+  Value reverse_index_map = sparse_fill_empty_rows_op.reverse_index_map();
+
+  Value num_rows = b.create<arith::IndexCastOp>(
+      loc, b.getIndexType(), b.create<memref::LoadOp>(loc, dense_shape, zero));
+  // Init all to true
+  {
+    auto for_op = b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                                       /* upperBound */ num_rows,
+                                       /* step */ one);
+    for_op.getBody()->clear();
+    b.setInsertionPointToStart(for_op.getBody());
+
+    Value i = for_op.getInductionVar();
+    b.create<memref::StoreOp>(loc, true_value, empty_row_indicator, i);
+    b.create<scf::YieldOp>(loc, ValueRange({}));
+
+    b.setInsertionPointAfter(for_op);
+  }
+  // Init all to 0
+  Value row_count_memref;
+  {
+    auto alloc = b.create<memref::AllocOp>(
+        loc, MemRefType::get({-1}, b.getIntegerType(64)), num_rows);
+    row_count_memref = alloc.getResult();
+    // b.s:setInsertionPointAfter(sparse_fill_empty_rows_op.getOperation());
+
+    auto for_op = b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                                       /* upperBound */ num_rows,
+                                       /* step */ one);
+    for_op.getBody()->clear();
+    b.setInsertionPointToStart(for_op.getBody());
+
+    Value i = for_op.getInductionVar();
+    b.create<memref::StoreOp>(loc, zero_int, row_count_memref, i);
+    b.create<scf::YieldOp>(loc, ValueRange({}));
+
+    b.setInsertionPointAfter(for_op);
+  }
```

Comment:
use a lambda or helper function to to eliminate duplicated `memset`-like code?

Reply:
done